### PR TITLE
Use CNI cache to find container IP

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis (OpenFaaS Ltd) <alexellis2@gmail.com>

## Description

Use CNI cache to find container IP

## Motivation and Context

This is an optimization that uses the results cache created by
CNI on the filesystem to store and fetch IP addresses for
containers in the core services and for functions. As part of
the change, the dependency on the syscall code from Weave net
has been removed, and the code should compile on MacOS again.

Updates and rebases the work in #38 by carlosedp

Tested in the original PR, further testing in the incoming
PR.

## How Has This Been Tested?

Tested end to end with the unit change and refactoring.

```
ubuntu@faasd:~$ sudo ls /var/lib/cni/results/
cni-loopback-basic-auth-plugin-3358-lo	openfaas-cni-bridge-basic-auth-plugin-3358-eth1
cni-loopback-gateway-3571-lo		openfaas-cni-bridge-gateway-3571-eth1
cni-loopback-nats-3120-lo		openfaas-cni-bridge-nats-3120-eth1
cni-loopback-prometheus-3464-lo		openfaas-cni-bridge-prometheus-3464-eth1
cni-loopback-queue-worker-3254-lo	openfaas-cni-bridge-queue-worker-3254-eth1
ubuntu@faasd:~$ sudo cat /var/lib/cni/results/cni-loopback-gateway-3571-lo
{"cniVersion":"0.3.1","interfaces":[{"name":"lo","mac":"00:00:00:00:00:00","sandbox":"/proc/3571/ns/net"}],"ips":[{"version":"4","interface":0,"address":"127.0.0.1/8"},{"version":"6","interface":0,"address":"::1/128"}],"dns":{}}ubuntu@faasd:~$ 
ubuntu@faasd:~$ 
ubuntu@faasd:~$ sudo cat /var/lib/cni/results/openfaas-cni-bridge-gateway-3571-eth1
{"cniVersion":"0.4.0","interfaces":[{"name":"openfaas0","mac":"3a:15:bb:8c:c9:dd"},{"name":"vethdc807b63","mac":"2e:69:fd:88:c9:33"},{"name":"eth1","mac":"96:60:e9:2a:ac:40","sandbox":"/proc/3571/ns/net"}],"ips":[{"version":"4","interface":2,"address":"10.62.0.6/16","gateway":"10.62.0.1"}],"routes":[{"dst":"0.0.0.0/0"}],"dns":{}}ubuntu@faasd:~$ 
ubuntu@faasd:~$ 
ubuntu@faasd:~$ curl 10.62.0.6:8080
<a href="/ui/">Moved Permanently</a>.

ubuntu@faasd:~$ 

```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

To give an idea of the impact of this change.

`faas-cli list` with 1024 containers was taking 15s on the most powerful 64-core / 384 GB RAM bare-metal host that Equinix Metal have

Today, on a 2GB / 2vCPU Multipass VM on my Intel NUC I'm seeing 0.6s for 200 functions.

### CODE REVIEW

File: pkg/cninetwork/cni_network.go [hunk 1]
```diff
@@ -131,43 +147,33 @@ func DeleteCNINetwork(ctx context.Context, cni gocni.CNI, client *containerd.Cli
...
 	}
 
-	if len(addrs) > 0 && len(addrs[0].CIDRs) > 0 {
-		return &addrs[0].CIDRs[0].IP, nil
+	for _, file := range files {
+		f, err := os.Open(filepath.Join(CNIDir, file.Name()))
+		if err != nil {
+			return "", fmt.Errorf("failed to open CNI IP file for %s/%s: %v", CNIDir, file.Name(), err)
+		}
+		reader := bufio.NewReader(f)
```
Comment (alexellis) [inline pkg/cninetwork/cni_network.go:165]: @carlosedp do you remember what this is doing?

I've added output of what the files look like in the PR text.

Comment (carlosedp) [reply pkg/cninetwork/cni_network.go:165]: I remember each IP allocation saves the container IP on a file in this dir...
Need to take a look and if needed rework this whole PR.

Comment (alexellis) [reply pkg/cninetwork/cni_network.go:165]: Look at the PR description if you like. I provided two samples for you to view without setting anything up.

Just curious about the approach chosen for parsing a JSON file.

Comment (alexellis) [reply pkg/cninetwork/cni_network.go:165]: It turns out this was a text file. I've documented it in a follow-up commit and have added a unit test for it. cc @LucasRoesler

**Pull request merged**